src/spark_rapids_ml/knn.py

```diff
@@ -0,0 +1,400 @@
+    StructType,
+)
+
+from spark_rapids_ml.core import (
+    INIT_PARAMETERS_NAME,
+    CumlInputType,
+    CumlT,
+    _CumlEstimatorSupervised,
+    _CumlModel,
+    _set_pyspark_cuml_cls_param_attrs,
+    alias,
+)
+from spark_rapids_ml.params import _CumlClass, _CumlParams
+
+
+class NearestNeighborsClass(_CumlClass):
+    @classmethod
+    def _cuml_cls(cls) -> List[type]:
+        from cuml import NearestNeighbors as cumlNearestNeighbors
+        from cuml.neighbors.nearest_neighbors_mg import (
+            NearestNeighborsMG,  # to include the batch_size parameter that exists in the MG class
+        )
+
+        return [cumlNearestNeighbors, NearestNeighborsMG]
+
+    @classmethod
+    def _param_mapping(cls) -> Dict[str, Optional[str]]:
+        return {"k": "n_neighbors"}
+
+    @classmethod
+    def _param_excludes(cls) -> List[str]:
+        return [
+            "handle",
+            "algorithm",
+            "metric",
+            "p",
+            "algo_params",
+            "metric_expanded",
+            "metric_params",
+            "output_type",
+        ]
+
+
+class _NearestNeighborsCumlParams(_CumlParams, HasInputCol, HasLabelCol):
+    """
+    Shared Spark Params for NearestNeighbor and NearestNeighborModel.
+    """
+
+    k = Param(
+        Params._dummy(),
+        "k",
+        "The number nearest neighbors to retrieve. Must be >= 1.",
+        typeConverter=TypeConverters.toInt,
+    )
+
+    id_col = Param(
+        Params._dummy(),
+        "id_col",
+        "id column name.",
+        typeConverter=TypeConverters.toString,
+    )
+
+    def setInputCol(self, value: str) -> "_NearestNeighborsCumlParams":
+        """
+        Sets the value of `inputCol`.
+        """
+        self.set_params(inputCol=value)
+        return self
+
+    def setIdCol(self, value: str) -> "_NearestNeighborsCumlParams":
+        """
+        Sets the value of `id_col`. If not set, zipWithIndex will be used and the row order is not guaranted to be deterministic.
+        """
+        self.set_params(id_col=value)
+        return self
+
+    def getIdCol(self) -> str:
+        """
+        Gets the value of `id_col`.
+        """
+        col_name = (
+            self.getOrDefault("id_col")
+            if self.isDefined("id_col")
+            else alias.row_number
+        )
+        return col_name
+
+
+class _CumlEstimatorNearestNeighbors(
+    _CumlEstimatorSupervised, _NearestNeighborsCumlParams
+):
+    """
+    Base class for Cuml Nearest Neighbor.
+    """
+
+    def _pre_process_data(
+        self, dataset: DataFrame
+    ) -> Tuple[
+        List[Column], Optional[List[str]], int, Union[Type[FloatType], Type[DoubleType]]
+    ]:
+        (
+            select_cols,
+            multi_col_names,
+            dimension,
+            feature_type,
+        ) = super()._pre_process_data(dataset)
+
+        if self.hasParam("id_col") and self.isDefined("id_col"):
+            label_name = self.getOrDefault("id_col")
```

Comment:
Why label_name as variable name here?

Reply:
Good catch. Revised the variable name to id_col_name.